### PR TITLE
Notify startup fail

### PR DESCRIPTION
**Pull request by seliayeu**
Added a cross-platform notification when Tauri fails to launch. Rather ugly on Windows since it looks like apps that aren't installed don't have a "AppUserModeId" which is what Windows seems to use identify the app behind the notification and add its name/icon to the banner. Currently just says "Windows Powershell". Also seems to support an additional type of icon as a cross-platform feature but couldn't get it working on Windows ☹️.

### CODE REVIEW

**Review by Ynng — APPROVED** [review]
lgtm

**Review by CheatCod — CHANGES_REQUESTED** [review]
Sorry for the confusion, but your PR should:
1. Make the `run` method in core return a `Result<..., Error>`
2. Match the result of running core on Tauri
3. If the result is Err then display a popup before shutting down the app

**Comment by Ynng** [comment]
Is this ready for review again?
And can we resolve the merge conflict

File: dashboard/src-tauri/src/main.rs [hunk 1]
```diff
@@ -106,17 +128,25 @@ async fn main() {
...
             }
-            _ => {}
-        })
-        .run(context)
-        .expect("error while running tauri application");
+                _ => {}
+            })
+            .run(context)
+    {
+        Ok(_) => {}
```
Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:142]: it would be nice if we can resolve some of these clippy warnings

File: core/src/lib.rs [hunk 2]
```diff
@@ -373,12 +373,12 @@ pub struct Args {
...
 pub async fn run(
     args: Args,
-) -> (
+) -> Result<(
     impl Future<Output = ()>,
     AppState,
     tracing_appender::non_blocking::WorkerGuard,
     tokio::sync::oneshot::Sender<()>,
-) {
+), Error> {
```
Comment (Ynng) [inline core/src/lib.rs:381]: I'm a little confused, I don't see any question mark in this function, where could Err(Error) come from?

**Review by Ynng — COMMENTED** [review]
I'll build it locally and try it when I get back

File: core/src/lib.rs (hunk 2 above)
Comment (seliayeu) [reply core/src/lib.rs:382]: yeah sorry i forgot about, i've updated `run` to actually throw Errors now

File: dashboard/src-tauri/src/main.rs [hunk 3]
```diff
@@ -50,12 +51,33 @@ async fn get_owner_jwt(state: tauri::State<'_, AppState>) -> Result<JwtToken, ()
...
+
+    let (core_fut, app_state, _guard, shutdown_tx);
+    match run_result {
+        Ok((fut, state, guard, tx)) => {
+            core_fut = fut;
+            app_state = state;
+            _guard = guard;
+            shutdown_tx = tx;
+        }
+        Err(_) => {
```
Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:69]: Might be good if we display the error message in the notification body
Ideally we can make it pretty

File: dashboard/src-tauri/src/main.rs [hunk 4]
```diff
@@ -81,7 +103,7 @@ async fn main() {
 
     let tray_menu = SystemTrayMenu::new().add_item(quit);
 
-    builder
+    if let Err(_) = builder
```
Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:106]: display error message in the notification body

File: dashboard/src-tauri/src/main.rs [hunk 5]
```diff
@@ -106,17 +128,22 @@ async fn main() {
                     if let Some(tx) = shutdown_tx.lock().unwrap().take() {
                         tx.send(()).unwrap();
                     }
```
Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:130]: Why are we removing this app.exit(0)?

**Review by Ynng — APPROVED** [review]
Works on my machine, just some minor things

**Review by CheatCod — APPROVED** [review]

File: dashboard/src-tauri/src/main.rs (hunk 5 above)
Comment (seliayeu) [reply dashboard/src-tauri/src/main.rs:130]: i think i accidentally deleted that in a merge, added it back now

File: dashboard/src-tauri/src/main.rs (hunk 3 above)
Comment (seliayeu) [reply dashboard/src-tauri/src/main.rs:69]: added. wasn't sure if that's relevant information for the end user so i reserved from adding it before

File: dashboard/src-tauri/src/main.rs (hunk 4 above)
Comment (seliayeu) [reply dashboard/src-tauri/src/main.rs:106]: added. this one isn't our Error though so not 100% sure if it will be pretty

**Review by Ynng — APPROVED** [review]

File: dashboard/src-tauri/src/main.rs [hunk 6]
```diff
@@ -68,7 +90,7 @@ async fn main() {
     #[cfg(not(dev))]
     {
         let port = portpicker::pick_unused_port().expect("Failed to pick unused port");
-        let url = format!("http://localhost:{}", port).parse().unwrap();
+        let url = format!(&"http://localhost:{}", port).parse().unwrap();
```
Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:93]: I think the & might be causing the clippy error?

Comment (Ynng) [inline dashboard/src-tauri/src/main.rs:93]:
```suggestion
        let url = format!("http://localhost:{port}").parse().unwrap();
```

Comment (Ynng) [reply dashboard/src-tauri/src/main.rs:93]: This should fix the build error

**Review by Ynng — CHANGES_REQUESTED** [review]

**Review by Ynng — APPROVED** [review]
lgtm! tested both normal launch bad launch in fedora

**Pull request merged**